b2/sync.py

```diff
@@ -761,64 +762,123 @@ def check_file_replacement(source_file, dest_file, args):
         raise CommandError('Invalid option for --compareVersions')
 
 
+def should_transfer_file(source_file, dest_file, args):
+    """
+    Decides whether to transfer the file from the source to the destination.
+    """
+    if source_file is None:
+        # No source file.  Nothing to transfer.
+        return False
+    elif dest_file is None:
+        # Source file exists, but no destination file.  Always transfer.
+        return True
+    else:
+        # Both exist.  Transfer only if the two are different.
+        return files_are_different(source_file, dest_file, args)
+
+
+def make_b2_delete_actions(source_file, dest_file, dest_folder, transferred):
+    for (version_index, version) in enumerate(dest_file.versions):
+        keep = (version_index == 0) and (source_file is not None) and not transferred
```

Comment:
I think the inverted logic of this is much more readable: Delete a file it it's not the newest version, there is no source file or a new file is beeing transfered.